File: frontend/src/config/watchlist.ts
/**
 * Static watchlist for the executive grid dashboard.
 *
 * Three tiers, four tickers each, total 12 — the spec calls for a trader to
 * scan all 12 in under 30 seconds. The list is static on purpose: it ships
 * as a compile-time config, no admin UI required for v1. Re-tier or expand
 * by editing this file directly.
 */

export type WatchlistTicker = {
  symbol: string;
  company: string;
};

export type WatchlistTier = {
  key: string;
  name: string;
  description?: string;
  tickers: WatchlistTicker[];
};

export const WATCHLIST_TIERS: WatchlistTier[] = [
  {
    key: "tier-1",
    name: "Tier 1",
    description: "Index ETFs",
    tickers: [
      { symbol: "SPY", company: "SPDR S&P 500 ETF Trust" },
      { symbol: "QQQ", company: "Invesco QQQ Trust" },
      { symbol: "IWM", company: "iShares Russell 2000 ETF" },
      { symbol: "DIA", company: "SPDR Dow Jones Industrial Average ETF" },
    ],
  },
  {
    key: "tier-2",
    name: "Tier 2",
    description: "Mega-cap tech",
    tickers: [
      { symbol: "AAPL", company: "Apple Inc." },
      { symbol: "MSFT", company: "Microsoft Corporation" },
      { symbol: "AMZN", company: "Amazon.com, Inc." },
      { symbol: "GOOGL", company: "Alphabet Inc." },
    ],
  },
  {
    key: "tier-3",
    name: "Tier 3",
    description: "High-beta growth",
    tickers: [
      { symbol: "NVDA", company: "NVIDIA Corporation" },
      { symbol: "TSLA", company: "Tesla, Inc." },
      { symbol: "AMD", company: "Advanced Micro Devices, Inc." },
      { symbol: "META", company: "Meta Platforms, Inc." },
    ],
  },
];

export const ALL_WATCHLIST_TICKERS: string[] = WATCHLIST_TIERS.flatMap((t) =>
  t.tickers.map((s) => s.symbol),
);

export const WATCHLIST_COMPANY_BY_SYMBOL: Record<string, string> = Object.fromEntries(
  WATCHLIST_TIERS.flatMap((t) => t.tickers.map((s) => [s.symbol, s.company])),
);

File: backend/app/services/dashboard/watchlist.py
"""Static watchlist tier definitions for the Reverse BWB dashboard.

This is the authoritative server-side mirror of
``frontend/src/config/watchlist.ts``. Both files must move together when the
watchlist is rebalanced. The order of ``WATCHLIST_TIERS`` defines the
sequential execution order enforced by ``WatchlistBatchService``:

    SPY -> QQQ -> IWM -> DIA -> AAPL -> MSFT -> AMZN -> GOOGL
        -> NVDA -> TSLA -> AMD -> META
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class WatchlistTicker:
    symbol: str
    company: str


@dataclass(frozen=True)
class WatchlistTier:
    key: str
    name: str
    description: str
    tickers: tuple[WatchlistTicker, ...]


WATCHLIST_TIERS: tuple[WatchlistTier, ...] = (
    WatchlistTier(
        key="tier-1",
        name="Tier 1",
        description="Index ETFs",
        tickers=(
            WatchlistTicker("SPY", "SPDR S&P 500 ETF Trust"),
            WatchlistTicker("QQQ", "Invesco QQQ Trust"),
            WatchlistTicker("IWM", "iShares Russell 2000 ETF"),
            WatchlistTicker("DIA", "SPDR Dow Jones Industrial Average ETF"),
        ),
    ),
    WatchlistTier(
        key="tier-2",
        name="Tier 2",
        description="Mega-cap tech",
        tickers=(
            WatchlistTicker("AAPL", "Apple Inc."),
            WatchlistTicker("MSFT", "Microsoft Corporation"),
            WatchlistTicker("AMZN", "Amazon.com, Inc."),
            WatchlistTicker("GOOGL", "Alphabet Inc."),
        ),
    ),
    WatchlistTier(
        key="tier-3",
        name="Tier 3",
        description="High-beta growth",
        tickers=(
            WatchlistTicker("NVDA", "NVIDIA Corporation"),
            WatchlistTicker("TSLA", "Tesla, Inc."),
            WatchlistTicker("AMD", "Advanced Micro Devices, Inc."),
            WatchlistTicker("META", "Meta Platforms, Inc."),
        ),
    ),
)


ALL_WATCHLIST_TICKERS: tuple[str, ...] = tuple(
    t.symbol for tier in WATCHLIST_TIERS for t in tier.tickers
)


WATCHLIST_COMPANY_BY_SYMBOL: dict[str, str] = {
    t.symbol: t.company for tier in WATCHLIST_TIERS for t in tier.tickers
}


WATCHLIST_TIER_KEY_BY_SYMBOL: dict[str, str] = {
    t.symbol: tier.key for tier in WATCHLIST_TIERS for t in tier.tickers
}


def is_watchlist_ticker(ticker: str) -> bool:
    return ticker.upper() in WATCHLIST_COMPANY_BY_SYMBOL


def tier_for(ticker: str) -> str | None:
    return WATCHLIST_TIER_KEY_BY_SYMBOL.get(ticker.upper())
